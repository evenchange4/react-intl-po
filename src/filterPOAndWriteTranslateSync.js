/* eslint-disable no-console */
import fs from 'fs';
import path from 'path';
import mkdirp from 'mkdirp';
import chalk from 'chalk';
import values from 'lodash/values';
import flatten from 'lodash/flatten';
import flowRight from 'lodash/flowRight';
import toObjectBy from 'to-object-by';
import readAllMessageAsObjectSync from './readAllMessageAsObjectSync';
import readAllPOAsObjectSync from './readAllPOAsObjectSync';

function filterPOAndWriteTranslateSync(srcPatterns, { messagesPattern, output }) {
  const translationTable = readAllPOAsObjectSync(srcPatterns);
  const messageList = flowRight(
    flatten,                    // 3. return flatten object values
    values,                     // 2. return object values
    readAllMessageAsObjectSync, // 1. return message object
  )(messagesPattern);

  const locales = Object.keys(translationTable);
  const result = toObjectBy(locales, locale => ({
    [locale]: toObjectBy(messageList, ({ id, defaultMessage }) => ({
      [id]: translationTable[locale][defaultMessage],
    })),
  }));

  mkdirp(path.dirname(output)); // ensure the output folder exists
  fs.writeFileSync(output, JSON.stringify(result, null, 0));
  console.log(chalk.green(`> [react-intl-po] write file -> ${output} ✔️\n`));
}

export default filterPOAndWriteTranslateSync;
