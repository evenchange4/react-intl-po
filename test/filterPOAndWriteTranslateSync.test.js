import fs from 'fs';
import test from 'ava';
import filterPOAndWriteTranslateSync
  from '../src/filterPOAndWriteTranslateSync';

test('should return a function', t => {
  t.is(typeof filterPOAndWriteTranslateSync, 'function');
});

test('should return messages object with default mapper', t => {
  const messagesPattern = './messages/**/*.json';
  const output = './translations.json';

  filterPOAndWriteTranslateSync('./po/mcs-public.*.po', { messagesPattern, output });
  t.deepEqual(
    JSON.parse(fs.readFileSync(output, 'utf8')),
    {
      'zh-CN': {
        'App.Creator': '建立者（簡中）',
        'NotFound.Creator': '建立者（簡中）',
      },
      'zh-TW': {
        'App.Creator': '建立者',
        'NotFound.Creator': '建立者',
      },
    },
  );
});

test('should return different objects if multifile option is enabled', t => {
  const messagesPattern = './messages/**/*.json';
  const output = './translations';
  const multi = true;

  filterPOAndWriteTranslateSync('./po/mcs-public.*.po', { messagesPattern, output, multi });
  t.deepEqual(
    JSON.parse(fs.readFileSync(`${output}/zh-CN.json`, 'utf8')),
    {
      'App.Creator': '建立者（簡中）',
      'NotFound.Creator': '建立者（簡中）',
    }
  );

  t.deepEqual(
    JSON.parse(fs.readFileSync(`${output}/zh-TW.json`, 'utf8')),
    {
      'App.Creator': '建立者',
      'NotFound.Creator': '建立者',
    }
  );
});
