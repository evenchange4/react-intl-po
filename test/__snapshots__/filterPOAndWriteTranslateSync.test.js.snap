// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should output correct filter merged file with id as messageContext 1`] = `
{
  "zh-CN": {
    "App.errorMessage": "",
    "App.errorButton": "",
    "NotFound.errorButton": "",
    "App.Creator 1": "建立者（簡中1）",
    "App.Creator 2": "建立者（簡中2）",
    "NotFound.Creator": ""
  },
  "zh-TW": {
    "App.errorMessage": "",
    "App.errorButton": "",
    "NotFound.errorButton": "",
    "App.Creator 1": "建立者（繁中1）",
    "App.Creator 2": "建立者（繁中2）",
    "NotFound.Creator": ""
  }
}
`;

exports[`should output correct filter merged file with id as messageKey 1`] = `
{
  "zh-CN": {
    "App.errorMessage": "",
    "App.errorButton": "",
    "App.Creator 1": "建立者（簡中1）",
    "App.Creator 2": "建立者（簡中2）",
    "NotFound.errorButton": "",
    "NotFound.Creator": ""
  },
  "zh-TW": {
    "App.errorMessage": "",
    "App.errorButton": "",
    "App.Creator 1": "建立者（繁中1）",
    "App.Creator 2": "建立者（繁中2）",
    "NotFound.errorButton": "",
    "NotFound.Creator": ""
  }
}
`;

exports[`should output one file per locale if a *directory* is set 1`] = `
{
  "App.errorMessage": "",
  "App.errorButton": "",
  "NotFound.errorButton": "",
  "App.Creator 1": "建立者（簡中）",
  "App.Creator 2": "建立者（簡中）",
  "NotFound.Creator": "建立者（簡中）"
}
`;

exports[`should output one file per locale if a *directory* is set 2`] = `
{
  "App.errorMessage": "",
  "App.errorButton": "",
  "NotFound.errorButton": "",
  "App.Creator 1": "建立者",
  "App.Creator 2": "建立者",
  "NotFound.Creator": "建立者"
}
`;

exports[`should output one merged file if a *json file* is set 1`] = `
{
  "zh-CN": {
    "App.errorMessage": "",
    "App.errorButton": "",
    "NotFound.errorButton": "",
    "App.Creator 1": "建立者（簡中）",
    "App.Creator 2": "建立者（簡中）",
    "NotFound.Creator": "建立者（簡中）"
  },
  "zh-TW": {
    "App.errorMessage": "",
    "App.errorButton": "",
    "NotFound.errorButton": "",
    "App.Creator 1": "建立者",
    "App.Creator 2": "建立者",
    "NotFound.Creator": "建立者"
  }
}
`;
